### PR TITLE
Deploy linchpin controller

### PR DESCRIPTION
**Pull request by rvykydal**

### CODE REVIEW

File: ansible/roles/kstest-controller/vars/main.yml [hunk 1]
```diff
@@ -1,17 +1,6 @@
...
-
-kstest_controller_user: kstest
-linchpin_user: kstest
-private_keys_to_upload:
-  - "/home/rvykydal/.ssh/kstests.pem"
-public_keys_to_upload:
-  - "/home/rvykydal/.ssh/kstests.pub"
-  - "/home/rvykydal/.ssh/id_rsa.pub"
-cloud_config_file: /home/rvykydal/.config/linchpin/clouds.yml
-
```
Comment (jkonecny12) [inline ansible/roles/kstest-controller/vars/main.yml]: Please remove this from a history completely.

File: linchpin/README.md [hunk 2]
```diff
@@ -10,7 +10,21 @@ To pull *runners* provisioning (linchpin) and deployment (ansible), *test run* c
 Requirements
 -----------
 
-The script requires `linchpin` and `ansible` to be installed. Linchpin `pip` installation in `virtualenv` is desribed in [documentation](https://linchpin.readthedocs.io).
+The script requires `linchpin` and `ansible` to be installed. Linchpin `pip` installation in `virtualenv` is desribed in [documentation](https://linchpin.readthedocs.io).
```
Comment (jkonecny12) [inline linchpin/README.md]: typo:
`desribed` -> `described`

File: linchpin/README.md [hunk 3]
```diff
@@ -124,7 +138,20 @@ It is also possible to set up the scheduling on the *master* of a permanent *tar
 Examples
 --------
 
-[examples](examples)
+[Testing a code change running a single test on temporary target](examples/example1)
```
Comment (jkonecny12) [inline linchpin/README.md]: On other places you have `Testing a change...` please make that consistent.

File: linchpin/examples/controller-provisioning-linchpin/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+In this example the variables can be configured in the provisioning script [kstests-controller-provision.sh](kstests-controller-provision.sh) by updating the `[kstest-controller:vars]` section of the generated inventory.
+
+## Provision and deploy:
+
+Copy [PinFile.controller](linchpin/PinFile.controller) containing resources definition of *controller* host to the `linchpin` folder.
+
+Copy [kstests-controller-provision.sh](kstests-controller-provision.sh) in `kickstart-tests` repo root and set the path to the private key in `<PRIVATE_KEY_PATH>`.
+Also set credentials to be passed to the *controller* in `<PUBLIC_KEY_TO_UPLOAD_PATH>`, `<PRIVATE_KEY_TO_UPLOAD_PATH>`, `<CLOUD_CONFIG_FILE>` (or comment out the lines if you don't intend to use them on the controller).
+
+Run [kstests-controller-provision.sh](kstests-controller-provision.sh) in `kicstart-tests` repo root.
```
Comment (jkonecny12) [inline linchpin/examples/controller-provisioning-linchpin/README.md]: Typo `kicstart-tests` -> `kickstart-tests`

File: linchpin/examples/controller-provisioning/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+## Credentials configuration:
+
+### Provisioning:
+
+Kickstart to be used for controller installation: [ks.kstest-controller.cfg](ks.kstest-controller.cfg)
+
+- public ssh key to access the host is passed in `sshkey` command
+
+Inventory for the deployment playbook: [hosts](hosts)
```
Comment (jkonecny12) [inline linchpin/examples/controller-provisioning/README.md]: The `(hosts)` should be a path so please change this to `./hosts`.

File: linchpin/examples/controller-provisioning/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+In this example the variables can be configured in the `[kstest-controller:vars]` section of the inventory [hosts](hosts).
+
+## Provision and deploy:
+
+Modify ssh key configuration in the [ks.kstest-controller.cfg](ks.kstest-controller.cfg) for your keypair.
+
+Install the controller host using the [ks.kstest-controller.cfg](ks.kstest-controller.cfg) kickstart.
+
+Copy the inventory [hosts](hosts) into `kickstart-tests` root and replace the <IP_ADDRESS> with the real controller IP address.
```
Comment (jkonecny12) [inline linchpin/examples/controller-provisioning/README.md]: The same here.

File: linchpin/examples/controller-provisioning/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+Update the credentials configuration in [hosts](hosts) - set the path to the (deployment) private key in `<PRIVATE_KEY_PATH>`. Also set credentials to be passed to the *controller* in `<PUBLIC_KEY_TO_UPLOAD_PATH>`, `<PRIVATE_KEY_TO_UPLOAD_PATH>`, `<CLOUD_CONFIG_FILE>` (or comment out the lines if you don't intend to use them on the controller).
```
Comment (jkonecny12) [inline linchpin/examples/controller-provisioning/README.md]: and here

File: linchpin/examples/controller-provisioning/ks.kstest-controller.cfg [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+clearpart --all
+
+autopart
+
+%packages --ignoremissing
+@core
+@base
+%end
+
+sshkey --username root "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDJjCafb/FZ4hIuQXfUPvhkZmQT2na//fl2rmnG5i65EyJXrWZyO2VM4n4rDmYT2scxp3BDV1ovacTetiqUW3cBID+pzuyVmRonsE6eNudlP47BWv14kPY3SocImt4lgjA2bag4yhgN7mx20c3bLGL7L25D27SZwDYfBbPUl1uuypmMlyiLZ7Cwpf+qEeobfk5nwjZMDZEL9abPk2HYIm+2udLCqUtVj/51lQ+XL5AH1EJvMdh4Fv6GuqV69jSZoc9XcPoUX22NXIirym0Tpzro/j4xOZ0jDN5E012fdMcKwFwLMiXrSq1Amgk4B7z0EQQ4z3/by/EAsxCEkyuoa5vr Generated-by-Nova"
```
Comment (jkonecny12) [inline linchpin/examples/controller-provisioning/ks.kstest-controller.cfg]: What is this key, is it real key used somewhere?

File: ansible/README.md [hunk 9]
```diff
@@ -13,17 +13,20 @@ The playbooks
...
+* [`kstest-master-configure-test.yml`](kstest-master-configure-test.yml) - [configures a *test run*](#test-configuration) to be run from *master* on *runners*
+* [`kstest-master-run-test.yml`](kstest-master-run-test.yml) - runs a *test run* from *master* on *runners* (including the *master*)
+* [`kstest-master-show-test-status.yml`](kstest-master-show-test-status.yml) - shows the status of a *test run* initatiated from *master*
+* [`kstest-master-fetch-results.yml`](kstest-master-fetch-results.yml) - synchronizes test [results](#results) from *master* to local host
+* [`kstest-master-schedule-test.yml`](kstest-master-schedule-test.yml) - [schedules](#scheduling-configuration) *test runs* on *master*
 
+*controller*:
+
+* [`kstest-controller-deploy.yml`](kstest-controller-deploy.yml) - deploys [*controller*](../linchpin/README.md#controller) - a host from which *test run* resources (*master* and *runner*) can be provisioned and tests run and scheduled using [linchpin tooling](../linchpin).
```
Comment (jkonecny12) [inline ansible/README.md]: Could you please rephrase this. It's difficult to follow the sentence.

File: ansible/roles/kstest-controller/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,171 @@
+---
+
+### Install required packages
+
+- name: Install packages
+  dnf:
+    name:
+      - git
+      - vim
+      - mc
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply ansible/roles/kstest-controller/tasks/main.yml]: I want it but I will remove it :)

File: ansible/roles/kstest-controller/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+  copy:
+    src: files/sudoers.d.lecture
+    dest: /etc/sudoers.d/lecture
+    owner: root
+    group: root
+    mode: 0440
+
+- name: Make sure there is a 'wheel' group
+  group:
+    name: whell
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply ansible/roles/kstest-controller/tasks/main.yml]: Good catch! I'll have to fix it also in kstest role

File: ansible/roles/kstest-controller/tasks/main.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (rvykydal) [reply ansible/roles/kstest-controller/tasks/main.yml]: Removed

File: ansible/roles/kstest-controller/tasks/main.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (rvykydal) [reply ansible/roles/kstest-controller/tasks/main.yml]: Fixed

File: ansible/roles/kstest-controller/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+  dnf:
+    name: "*"
+    state: latest
+
+- name: Install kernel-modules (reboot if updated)
+  dnf:
+    name: "kernel-modules"
+    state: present
+  notify:
+    - Reboot machine
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply ansible/roles/kstest-controller/tasks/main.yml]: Maybe sharing the handler makes sense, but the other 2 tasks I'd keep in the roles. The handler is used also by one more task in kstests role which I don't want to make depend on the other 2 tasks via the new role.

File: ansible/roles/kstest-controller/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+  become_user: "{{ linchpin_user }}"
+  pip:
+    name: linchpin>1.6.0
+    virtualenv: "{{ virtualenv_path }}"
+    virtualenv_python: "{{ virtualenv_python }}"
+
+- name: Selinux workaround (symlinks to system libraries)
+  become: yes
+  become_user: "{{ linchpin_user }}"
+  block:
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply ansible/roles/kstest-controller/tasks/main.yml]: Fixed

File: ansible/roles/kstest-controller/vars/main.yml (hunk 1 above)
Comment (rvykydal) [reply ansible/roles/kstest-controller/vars/main.yml]: Removed

File: linchpin/README.md (hunk 2 above)
Comment (rvykydal) [reply linchpin/README.md]: Fixed

File: linchpin/README.md (hunk 3 above)
Comment (rvykydal) [reply linchpin/README.md]: Made consistent.

File: linchpin/examples/controller-provisioning-linchpin/README.md (hunk 4 above)
Comment (rvykydal) [reply linchpin/examples/controller-provisioning-linchpin/README.md]: Fixed.

File: linchpin/examples/controller-provisioning/README.md (hunk 5 above)
Comment (rvykydal) [reply linchpin/examples/controller-provisioning/README.md]: `hosts` is a relative path, no?
I am using relative paths to a file like this in many other links.

File: linchpin/examples/controller-provisioning/README.md (hunk 6 above)
Comment (rvykydal) [reply linchpin/examples/controller-provisioning/README.md]: Same here

File: linchpin/examples/controller-provisioning/README.md (hunk 7 above)
Comment (rvykydal) [reply linchpin/examples/controller-provisioning/README.md]: and here

File: linchpin/examples/controller-provisioning/ks.kstest-controller.cfg (hunk 8 above)
Comment (rvykydal) [reply linchpin/examples/controller-provisioning/ks.kstest-controller.cfg]: It is kstests public key I am using. I can put there any other generated public key, but I don't think it matters.

File: ansible/README.md (hunk 9 above)
Comment (rvykydal) [reply ansible/README.md]: I tried to, I hope that now it is more clear (mind the link to *controller*, there is a picture).

File: ansible/roles/kstest-controller/tasks/main.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (jkonecny12) [reply ansible/roles/kstest-controller/tasks/main.yml]: OK, having to share the handlers make this a difficult.

File: linchpin/examples/controller-provisioning/README.md (hunk 5 above)
Comment (jkonecny12) [reply linchpin/examples/controller-provisioning/README.md]: OK in that case you can leave it this way. It's only a more obvious that it is a relative path with the `./` but I'm fine even with this.

File: linchpin/examples/controller-provisioning/ks.kstest-controller.cfg (hunk 8 above)
Comment (jkonecny12) [reply linchpin/examples/controller-provisioning/ks.kstest-controller.cfg]: I guess public key is fine.

**Pull request merged**